tensorflow-training/src/main/java/org/tensorflow/training/examples/MNISTTest.java

```diff
@@ -0,0 +1,331 @@
+    // Initialises the parameters.
+    session.runner().addTarget(INIT).run();
+    logger.info("Initialised the model parameters");
+
+    float[][][][] featureBatch = new float[minibatchSize][][][];
+    int[] labelBatch = new int[minibatchSize];
+
+    int interval = 0;
+    for (int i = 0; i < epochs; i++) {
+      logger.log(Level.INFO, "Starting epoch " + i);
+      //Tensor<?> epoch = Tensor.create(i);
+      for (int j = 0; j < data.length; j += minibatchSize) {
+        for (int k = j, m = 0; k < (j + minibatchSize) && k < data.length; k++, m++) {
+          featureBatch[m] = data[k];
+          labelBatch[m] = labels[k];
+        }
+        //logger.info("Batch = " + batch.size());
+        Tensor<?> input = Tensor.create(featureBatch);
+        Tensor<?> target = Tensor.create(labelBatch);
+        Tensor<?> loss = session.runner()
+            .feed(INPUT_NAME, input)
+            .feed(TARGET, target)
+            .addTarget(TRAIN)
+            .fetch(TRAINING_LOSS)
+            .run().get(0);
+        if (interval % 100 == 0) {
+          logger.log(Level.INFO, "Iteration = " + interval + ", training loss = " + loss.floatValue());
+        }
+        input.close();
+        target.close();
+        loss.close();
+        interval++;
+      }
+      //epoch.close();
+    }
+  }
+
+  /**
+   * Find the maximum probability and return it's index.
+   *
+   * @param probabilities The probabilites.
+   * @return The index of the max.
+   */
+  public static int pred(float[] probabilities) {
+    float maxVal = Float.NEGATIVE_INFINITY;
+    int idx = 0;
+    for (int i = 0; i < probabilities.length; i++) {
+      if (probabilities[i] > maxVal) {
+        maxVal = probabilities[i];
+        idx = i;
+      }
+    }
+    return idx;
+  }
+
+  public static DataTuple loadData(String path) throws IOException, ClassNotFoundException {
+    try (ObjectInputStream ois = new ObjectInputStream(new BufferedInputStream(new FileInputStream(path)))) {
+      float[][][][] data = (float[][][][]) ois.readObject();
+      int[] labels = (int[]) ois.readObject();
+      return new DataTuple(data, labels);
+    }
+  }
+
+  private static class DataTuple {
+    public final float[][][][] features;
+    public final int[] labels;
+
+    public DataTuple(float[][][][] features, int[] labels) {
+      this.features = features;
+      this.labels = labels;
+    }
+  }
+
+  public static void main(String[] args) throws IOException, ClassNotFoundException {
+    logger.info("Usage: MNISTTest <num-epochs> <minibatch-size> <optimizer-name> <train-data-path> <test-data-path>");
+
+    logger.info("Loading training data");
+    DataTuple train = loadData(args[3]);
+    logger.info("Loading testing data");
+    DataTuple test = loadData(args[4]);
+
+    logger.info("Loaded data.");
+
+    float[][][][] trainData = train.features;
+    int[] trainLabels = train.labels;
+
+    float[][][][] testData = test.features;
```

Comment:
That's because I load in multidimensional arrays from the serialised files I have lying around with MNIST in. When we move this over to the java-models repo I can use your MNIST loader which will get rid of this nonsense.

Reply:
Moved the file over to examples.